### ISSUE TITLE
S2583,S2589 FP: with nullable struct - update symbolic execution engine

### ISSUE THREAD

**Issue by andrei-epure-sonarsource** [SonarSource/sonar-dotnet#2496]
### Description

FP for rule [Boolean expressions should not be gratuitous](https://rules.sonarsource.com/csharp/RSPEC-2589) when a nullable struct like [Span](https://docs.microsoft.com/en-us/dotnet/api/system.span-1?view=netstandard-2.1) or `ReadOnlySpan` is null-checked inside an `if`.

### Repro steps

```
public void EmptyCollectionCheck(ReadOnlySpan<byte> span)
{
    // Check for empty collection with == override
    if (span == null) // FP S2589 Change this condition so it does not always evaluate to true
    {
        return;
    }
}
```

### Expected behavior

No issue

### Related information

* SonarC# Version 7.15

Initially reported [on community](https://community.sonarsource.com/t/c-ref-struct-span-and-readonlyspan-false-positive-as-sonarlint-understands-it-cannot-be-null/10090)

**Comment by abbas-sabra-sonarsource** [comment]
As discussed with @andrei-epure-sonarsource, this cannot be fixed without updating Roslyn version or adding a ShimLayer. We will go back to this ticket after one of the suggested fixes is put in place.

**Comment by andrei-epure-sonarsource** [comment]
cc @costin-zaharia-sonarsource @christophe-zurn-sonarsource

**Comment by andrei-epure-sonarsource** [comment]
Also [issue on peach](https://peach.sonarsource.com/issues?open=AW8fCP37MC6hVNLsGzZQ&projects=openiddict-core&resolved=false&rules=csharpsquid%3AS2583) for S2583

![image](https://user-images.githubusercontent.com/38876598/71251626-259b8c80-2323-11ea-8acd-62e3b2520e88.png)

[JsonElement](https://docs.microsoft.com/en-us/dotnet/api/system.text.json.jsonelement?view=netcore-3.1 )is a struct

**Comment by mary-georgiou-sonarsource** [comment]
Fixed by https://github.com/SonarSource/sonar-dotnet/pull/7750
